EssentialFeed/EssentialFeedTests/Feed API/LoadFeedCommentsFromRemoteUseCaseTests.swift

```diff
@@ -0,0 +1,192 @@
+		
+		_ = sut.load() { _ in }
+		_ = sut.load() { _ in }
+		
+		assert(requestedUrls: client.requestedURLs, equalsToExpectedUrls: [url, url])
+	}
+	
+	func test_load_deliversErrorOnClientError() {
+		let (sut, client, _) = makeSUT()
+		
+		expect(sut, toCompleteWith: failure(.connectivity), when: {
+			let clientError = NSError(domain: "Test", code: 0)
+			client.complete(with: clientError)
+		})
+	}
+	
+	func test_load_deliversErrorOnNon2xxHTTPResponse() {
+		let (sut, client, _) = makeSUT()
+		
+		let samples = [199, 300, 400, 500]
+		
+		samples.enumerated().forEach { index, code in
+			expect(sut, toCompleteWith: failure(.invalidData), when: {
+				let json = makeItemsJSON([])
+				client.complete(withStatusCode: code, data: json, at: index)
+			})
+		}
+	}
+	
+	func test_load_deliversErrorOn2xxHTTPResponseWithInvalidJSON() {
+		let (sut, client, _) = makeSUT()
+		
+		let samples = [200, 205, 206, 290, 299, 289]
+
+		samples.enumerated().forEach { index, code in
+			expect(sut, toCompleteWith: failure(.invalidData), when: {
+				let invalidJSON = Data("invalid json".utf8)
+				client.complete(withStatusCode: code, data: invalidJSON, at: index)
+			})
+		}
+	}
+	
+	func test_load_deliversNoItemsOn2xxCodeHTTPResponseWithEmptyJSONList() {
+		let (sut, client, _) = makeSUT()
+		
+		let samples = [200, 205, 206, 290, 299, 289]
+		
+		samples.enumerated().forEach { index, code in
+			expect(sut, toCompleteWith: .success([]), when: {
+				let emptyListJSON = makeItemsJSON([])
+				client.complete(withStatusCode: code, data: emptyListJSON, at: index)
+			})
+		}
+	}
+	
+	func test_load_deliversItemsOn2xxHTTPResponseWithJSONItems() {
+		let (sut, client, _) = makeSUT()
+		
+		let item1 = makeComment(
+			id: UUID(),
+			createdAt: (Date(timeIntervalSince1970: 1614618409), "2021-03-01T17:06:49+0000")
+		)
+		
+		let item2 = makeComment(
+			id: UUID(),
+			message: "test comment",
+			createdAt: (Date(timeIntervalSince1970: 0), "1970-01-01T00:00:00+0000")
+		)
+		
+		let items = [item1.model, item2.model]
+		
+		let samples = [200, 205, 206, 290, 299, 289]
+
+		samples.enumerated().forEach { index, code in
+			expect(sut, toCompleteWith: .success(items), when: {
+				let json = makeItemsJSON([item1.json, item2.json])
+				client.complete(withStatusCode: 200, data: json, at: index)
+			})
+		}
+
+	}
+	
+	func test_load_doesNotDeliverResultAfterSUTInstanceHasBeenDeallocated() {
+		let client = HTTPClientSpy()
+		
+		var sut: RemoteFeedImageCommentLoader? = RemoteFeedImageCommentLoader(url: anyURL(), client: client)
+		
+		var capturedResults = [RemoteFeedImageCommentLoader.Result]()
+		_ = sut?.load() { capturedResults.append($0) }
+		
+		sut = nil
+		client.complete(withStatusCode: 200, data: makeItemsJSON([]))
+		
+		XCTAssertTrue(capturedResults.isEmpty)
+	}
+	
+	private func makeSUT(url: URL = anyURL(), file: StaticString = #filePath, line: UInt = #line) -> (RemoteFeedImageCommentLoader, HTTPClientSpy, URL) {
+		let client = HTTPClientSpy()
```

Comment:
The `url: URL` parameter is never passed to this method, so it shouldn't be a parameter.

Create it inside the method implementation.

```suggestion
	private func makeSUT(file: StaticString = #filePath, line: UInt = #line) -> (RemoteFeedImageCommentLoader, HTTPClientSpy, URL) {
		let url = anyURL()
		let client = HTTPClientSpy()
```